### PR TITLE
getBoundsZoom with small size and padding

### PR DESCRIPTION
**Pull request by dianjin**
https://github.com/Leaflet/Leaflet/issues/4528

### CODE REVIEW

**Comment by dianjin** [comment]
@hyperknot: I tested with different zoomSnaps (with L.browser.any3d = true) and the returned value is always a multiple of the zoomSnap. Setting zoomSnap to zero looks like it's returning the most precise zoom level, in decimal. Do we want tests for this?

File: spec/suites/map/MapSpec.js [hunk 1]
```diff
@@ -222,27 +236,6 @@ describe("Map", function () {
...
-
-			L.tileLayer("{z}{x}{y}", {minZoom: 4, maxZoom: 10}).addTo(map);
-			L.tileLayer("{z}{x}{y}", {minZoom: 6, maxZoom: 17}).addTo(map);
-			L.tileLayer("{z}{x}{y}", {minZoom: 0, maxZoom: 22}).addTo(map);
-
-			expect(map.getMinZoom()).to.be(2);
-			expect(map.getMaxZoom()).to.be(20);
-		});
-	});
-
```
Comment (jieter) [inline spec/suites/map/MapSpec.js]: Seems you accidentally removed this test?

Comment (dianjin) [inline spec/suites/map/MapSpec.js]: I removed it on purpose, this test is an exact copy of the one right above it. (Is there a reason to include the same test twice?)

Comment (jieter) [inline spec/suites/map/MapSpec.js]: Ah, I see 👍

**Comment by hyperknot** [comment]
@dianjin nice work! I only wrote testing different zoomSnap values as it is important to understand the logic and to visually make sure that it works correctly. I believe there is no need to write a new test for it, zoomSnap has it's own tests.

File: spec/suites/map/MapSpec.js [hunk 2]
```diff
@@ -162,6 +162,20 @@ describe("Map", function () {
...
+		it("returns high levels of zoom with small areas and big padding", function () {
+			var container = map.getContainer();
+			container.style.height = '400px';
+			document.body.appendChild(container);
+
+			var halfLength = 0.0001;
+			var bounds = [[-halfLength, -halfLength], [halfLength, halfLength]];
+			var padding = [100, 100];
+
+			expect(map.getBoundsZoom(bounds, false, padding)).to.be.greaterThan(10);
```
Comment (hyperknot) [inline spec/suites/map/MapSpec.js]: Maybe it's better to go with a specific value, not just `greaterThan(10)`?

Comment (IvanSanchez) [inline spec/suites/map/MapSpec.js]: I agree - this should be deterministic.

**Comment by IvanSanchez** [comment]
> I believe there is no need to write a new test for it, `zoomSnap` has it's own tests.

Personally I'd feel a bit safer if we had a test for just `getBoundZoom` with the defaults, and one for `getBoundsZoom`+`zoomSnap`.

**Comment by dianjin** [comment]
Changed the tests in response to your feedback.

File: spec/suites/map/MapSpec.js [hunk 3]
```diff
@@ -162,6 +162,31 @@ describe("Map", function () {
...
+			expect(map.getBoundsZoom(bounds, false, padding)).to.be.equal(21);
+		});
+
+		it("returns multiples of zoomSnap when zoomSnap > 0 on any3d browsers", function () {
+			var container = map.getContainer();
+			container.style.height = height;
+			document.body.appendChild(container);
+			L.Browser.any3d = true;
+			map.options.zoomSnap = 5;
+			expect(map.getBoundsZoom(bounds, false, padding) % map.options.zoomSnap).to.be.equal(0);
```
Comment (IvanSanchez) [inline spec/suites/map/MapSpec.js]: Why not `to.be.equal(20.5)`?

File: spec/suites/map/MapSpec.js [hunk 4]
```diff
@@ -162,6 +162,31 @@ describe("Map", function () {
...
+		var height = '400px';
+
+		it("returns high levels of zoom with small areas and big padding", function () {
+			var container = map.getContainer();
+			container.style.height = height;
+			document.body.appendChild(container);
+			expect(map.getBoundsZoom(bounds, false, padding)).to.be.equal(21);
+		});
+
+		it("returns multiples of zoomSnap when zoomSnap > 0 on any3d browsers", function () {
```
Comment (IvanSanchez) [inline spec/suites/map/MapSpec.js]: Should be `it.skipInPhantom("returns ...")`. See `spec/suites/map/anim/Map.ZoomAnimationSpec.js` and the end of `spec/SpecHelper.js`

Comment (jieter) [inline spec/suites/map/MapSpec.js]: Hmm, I don't see why that should be, this test manually sets `any3d` to `true` and doesn't rely on css animations, or does it?

Comment (IvanSanchez) [inline spec/suites/map/MapSpec.js]: @jieter Ha, you're right, didn't see that at first glance.

But won't this change the value of `L.Browser.any3d` for future tests? We might just be lucky about the order of the tests.

Comment (jieter) [inline spec/suites/map/MapSpec.js]: Yeah, might want to set it to the original value after the test is done.

Comment (dianjin) [inline spec/suites/map/MapSpec.js]: Some of the other tests are also setting/resetting L.Browser.any3d (looks like this is the only property in L.Browser being manipulated though). Is there a way to reset L.Browser to default values in beforeEach?

File: spec/suites/map/MapSpec.js (hunk 3 above)
Comment (dianjin) [inline spec/suites/map/MapSpec.js]: I compared the modulus to zero so the test would match the test description more, but I can see that this is basically accepting an infinite number of results so I've changed it to just compare the result to a number.

**Comment by IvanSanchez** [comment]
@dianjin Everything looks good!

I'll tweak the values for the unit tests afterwards, and we'll worry about resetting `L.Browser.any3d` afterwards if needed (after all, the `it.skipInPhantom()` trick is working perfectly so far).

And sorry for making so many comments on your code - I know this makes us seem unfriendly from the eyes of a newcomer, but that's how we usually work :-)

Again, thanks for jumping in and fixing this bug! I hope it was an instructive experience, and hope you will keep an eye on other leaflet bugs :-)

**Comment by dianjin** [comment]
Thanks @IvanSanchez for the constructive feedback! I'm glad you guys pointed out where I could improve, thank you Leaflet and @Charlotteis for encouraging beginners to contribute! 😄

**Comment by Moriphete** [comment]
> @dianjin Everything looks good!
> 
> I'll tweak the values for the unit tests afterwards, and we'll worry about resetting `L.Browser.any3d` afterwards if needed (after all, the `it.skipInPhantom()` trick is working perfectly so far).
> 
> And sorry for making so many comments on your code - I know this makes us seem unfriendly from the eyes of a newcomer, but that's how we usually work :-)
> 
> Again, thanks for jumping in and fixing this bug! I hope it was an instructive experience, and hope you will keep an eye on other leaflet bugs :-)

Bugs Bats and unanswered queries. Now jow much time did it take to calculate the total depth of pie?

**Comment by Moriphete** [comment]
The total depth of pi

**Review by edwaka — COMMENTED** [review]
#4532

**Pull request merged**